src/Integration/Telemetry/TelemetryManager.cs

```diff
@@ -67,92 +62,86 @@ public bool IsAnonymousDataShared
-        {
-            var lastUploadDate = this.telemetryRepository.Data.LastUploadDate;
+        public bool IsAnonymousDataShared => this.telemetryRepository.Data.IsAnonymousDataShared;
 
-            if (lastUploadDate == DateTime.MinValue ||
-                IsNextDay(lastUploadDate))
-            {
-                this.telemetryRepository.Data.LastUploadDate = DateTime.Now;
-                this.telemetryRepository.Save();
+        public void Dispose()
+        {
+            DisableAllEvents();
 
-                await this.telemetryClient.SendPayload(GetPayload());
-            }
+            (this.tryUploadDataTimer as IDisposable)?.Dispose();
+            (this.firstCallDelayer as IDisposable)?.Dispose();
+            this.telemetryClient.Dispose();
+            this.telemetryRepository.Dispose();
         }
 
-        private bool IsNextDay(DateTime lastUploadDate)
+        public void OptIn()
         {
-            return (DateTime.Now - lastUploadDate).TotalDays >= 1 &&
-                (DateTime.Now - lastUploadDate).TotalHours >= MinHoursBetweenUpload;
+            this.telemetryRepository.Data.IsAnonymousDataShared = true;
+            this.telemetryRepository.Save();
+
+            EnableAllEvents();
         }
 
-        private void OnAnalysisRun(object sender, UIContextChangedEventArgs e)
+        public async void OptOut()
         {
-            if (!e.Activated)
-            {
-                return;
-            }
+            this.telemetryRepository.Data.IsAnonymousDataShared = false;
+            this.telemetryRepository.Save();
 
-            var lastAnalysisDate = this.telemetryRepository.Data.LastSavedAnalysisDate;
-            if (lastAnalysisDate == DateTime.MinValue ||
-                (DateTime.Now - lastAnalysisDate).TotalDays >= 1)
-            {
-                this.telemetryRepository.Data.LastSavedAnalysisDate = DateTime.Now;
-                this.telemetryRepository.Data.NumberOfDaysOfUse++;
-                this.telemetryRepository.Save();
-            }
+            DisableAllEvents();
+
+            await this.telemetryClient.OptOut(GetPayload());
         }
 
-        private TelemetryPayload GetPayload()
+        private static string GetSonarLintVersion()
         {
-            var numberOfDaysSinceInstallation = (long)(DateTime.Now - this.telemetryRepository.Data.InstallationDate).TotalDays;
-            return new TelemetryPayload
-            {
-                SonarLintProduct = "SonarLint Visual Studio",
-                SonarLintVersion = SonarLintVersion,
-                VisualStudioVersion = visualStudioVersion,
-                NumberOfDaysSinceInstallation = numberOfDaysSinceInstallation,
-                NumberOfDaysOfUse = this.telemetryRepository.Data.NumberOfDaysOfUse,
-                IsUsingConnectedMode = this.solutionBindingTracker.IsActiveSolutionBound
-            };
+            return FileVersionInfo.GetVersionInfo(Assembly.GetExecutingAssembly().Location).FileVersion;
```

Comment:
Apparently this works, but why not use typeof(SonarLintSomething).Assembly.Location? [MSDN says](https://msdn.microsoft.com/en-us/library/system.reflection.assembly.getexecutingassembly(v=vs.110).aspx) you should not call this method when you know what is the executing assembly...